<center>

$\Huge \textbf{Universidad Nacional Autónoma de México}$  
$\Huge \textbf{Facultad de Ciencias}$  
<p align="center">
  <img src="https://www.icat.unam.mx/wp-content/uploads/2021/11/Copia-de-LogoUNAM.-Azul.-Fondo-transparente.png" alt="UNAM" width="200"/>
</p>

<hr style="height:3px; background-color:#0B6E4F; border:none;"/>


$\LARGE \textbf{Inteligencia Artificial}$  

$\Large \textit{Laboratorio 5.8}$  


\begin{array}{rl}
\textbf{Docente:} & Dra. Jessica Sarahi Méndez Rincón \\[6pt]
\textbf{Ayudante de laboratorio:} & Diego Eduardo Peña Villegas \\[6pt]
\textbf{Alumna:} & Giovanni Alejandri Espinosa \\[6pt]
\textbf{Fecha de realización:} & 06/05/2026
\end{array}

</center>


**UNAM y la creadora del presente notebook, no guarda relación alguna con las marcas mencionadas como ejemplo. Las marcas son propiedad de sus titulares conforme a la legislación aplicable, se utilizan con fines académicos y didácticos, por lo que no existen fines de lucro, relación publicitaria o de patrocinio.

---

In [1]:
!pip install transformers

In [2]:
from transformers import pipeline, AutoTokenizer, AutoModelForSequenceClassification

# Cargar el modelo pre-entrenado y el tokenizador de BERT en español
#model_name = "dccuchile/bert-base-spanish-wwm-uncased"
# Este ya sabe qué es Positivo, Negativo y Neutral
# Muy robusto para español
model_name = "nlptown/bert-base-multilingual-uncased-sentiment"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

# Crear un pipeline para la clasificación de texto
text_classification_pipeline = pipeline(
    'text-classification',
    model=model,
    tokenizer=tokenizer
)

# Ejecutar la clasificación de texto en una oración en español
#result = text_classification_pipeline('Esta película es muy buena, la recomiendo totalmente')
#print(result)

# Ejecutar la clasificación
result = text_classification_pipeline('Esta película es muy buena, la recomiendo totalmente')

# Acceder a los datos dentro de la lista y el diccionario
label = result[0]['label']
score = result[0]['score']

print(f"Resultado: {label}")
print(f"Confianza: {score:.4f}")

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/953 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/669M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/201 [00:00<?, ?it/s]

Resultado: 5 stars
Confianza: 0.7252


1. El cambio en las etiquetas (Labels)
Al usar un modelo ajustado (probablemente el de nlptown o uno similar de reseñas), el sistema ya no usa códigos genéricos como LABEL_1.

"5 stars": El modelo ha clasificado la frase ("Esta película es muy buena, la recomiendo totalmente") con la puntuación máxima. Ha detectado un sentimiento altamente positivo.

2. La Confianza (Score)
0.7252 (72.5%): Es una métrica sólida. Indica que, de todas las opciones posibles (1, 2, 3, 4 o 5 estrellas), el modelo está bastante seguro de que pertenece a la categoría de 5.

In [3]:
from transformers import AutoTokenizer, AutoModelForCausalLM

# Descargar el modelo pre-entrenado de GPT-2 en español
model_name = "mrm8488/spanish-gpt2"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

config.json:   0%|          | 0.00/883 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/226 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

added_tokens.json:   0%|          | 0.00/24.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/90.0 [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/510M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/148 [00:00<?, ?it/s]

GPT2LMHeadModel LOAD REPORT from: mrm8488/spanish-gpt2
Key                                     | Status     |  | 
----------------------------------------+------------+--+-
transformer.h.{0...11}.attn.bias        | UNEXPECTED |  | 
transformer.h.{0...11}.attn.masked_bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [6]:
# Lista con los 3 ejemplos de texto inicial
input_strs = [
    "En el año 2050, la humanidad finalmente descubrió",
    "El detective entró a la habitación oscura y de repente",
    "El antiguo dragón despertó de su largo sueño al escuchar"
]

max_length = 100

# Generar texto con el modelo pre-entrenado para cada ejemplo
for i, input_str in enumerate(input_strs):
    print(f"--- GENERACIÓN DEL EJEMPLO {i+1} ---")
    print(f"Texto de entrada: '{input_str}'\n")

    input_ids = tokenizer.encode(input_str, return_tensors="pt")
    sample_outputs = model.generate(
        input_ids,
        do_sample=True,
        max_length=max_length,
        top_k=50,
        top_p=0.95,
        temperature=1.0,
        num_return_sequences=1
    )

    generated_text = tokenizer.decode(sample_outputs[0], skip_special_tokens=True)
    print(generated_text)
    print("\n" + "="*60 + "\n")

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


--- GENERACIÓN DEL EJEMPLO 1 ---
Texto de entrada: 'En el año 2050, la humanidad finalmente descubrió'



The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


En el año 2050, la humanidad finalmente descubrió sus verdaderos límites.En realidad no hay límite en el número de seres humanos de la Tierra.De hecho, los seres humanos tienen una única especie, el hombre.¿Quién?Yo.No puede ser.No hay forma de que no haya existido por al menos 100 millones de años.En la antigua era el hombre ha existido antes de la invención de la electricidad.El hombre primitivo no puede siquiera imaginar qué tan rápido el motor del crecimiento del árbol


--- GENERACIÓN DEL EJEMPLO 2 ---
Texto de entrada: 'El detective entró a la habitación oscura y de repente'



The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


El detective entró a la habitación oscura y de repente se estaba ahogando.Los paramédicos llegaron y lo pusieron en el piso.- El forense está en camino.- Bien.Y por lo menos estamos bien de nuevo.- Gracias, Dr. Harper.- Gracias.¿Quieres hablar de lo que pasó después?Sí, claro.¿Qué pasó?Bueno, hubo una explosión.El equipo de rescate encontró restos de cocaína.Lo vi en las noticias, y luego me di cuenta de que habían


--- GENERACIÓN DEL EJEMPLO 3 ---
Texto de entrada: 'El antiguo dragón despertó de su largo sueño al escuchar'

El antiguo dragón despertó de su largo sueño al escuchar el mensaje de la reina y se dirige a la ciudad.Allí, descubre que la reina está encerrada y que todos los prisioneros están dispuestos a morir por ella.Sin embargo, el dragón se libera de los que lo rodeaban y se dirige a los calabozos.Sin embargo, la reina es liberada en su propio mundo y su castillo fue destruido por el dragón, tras lo cual la reina es encerrada en un calabozo.Cuando el dragón es liberado,

In [5]:
# Lista con 3 ejemplos de texto inicial (incluyendo el que me pasaste)
input_strs = [
    "El hombre bicentenario es",
    "El androide miró sus manos mecánicas y de pronto pensó que",
    "Según las leyes de la robótica, una máquina nunca debe"
]

max_length = 100

# Generar texto con el modelo pre-entrenado de GPT-2 en español
for i, input_str in enumerate(input_strs):
    print(f"=== EJEMPLO {i+1} ===")
    print(f"Texto inicial: '{input_str}'\n")

    input_ids = tokenizer.encode(input_str, return_tensors="pt")

    # Generación
    sample_outputs = model.generate(
        input_ids,
        do_sample=True,
        max_length=max_length,
        top_k=50,
        top_p=0.95,
        temperature=1.0,
        num_return_sequences=1
    )

    # Decodificar y mostrar el resultado
    generated_text = tokenizer.decode(sample_outputs[0], skip_special_tokens=True)
    print("Texto generado por GPT-2:")
    print(generated_text)
    print("\n" + "="*60 + "\n")

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


El hombre bicentenario es un concepto muy amplio.No es meramente la ficción, sino que es una realidad tangible y es un concepto de verdad.No es un sueño.Es un evento.Un fenómeno.¿Qué es un bicentenario?Es un día especial.Es un día especial.Porque eres tú.¿De qué hablas?¿Yo soy qué?Eres tú.Estás loco.Yo, yo, nosotros.Eres tú.Eres nosotros.¡No es cierto!¡No es cierto!


## Conclusión

En este laboratorio, hemos explorado dos modelos fundamentales de procesamiento de lenguaje natural (NLP): BERT para análisis de sentimientos y GPT-2 para generación de texto. Hemos visto cómo BERT puede clasificar el sentimiento de una oración en español con un alto grado de confianza, demostrando su capacidad para comprender el contexto y la polaridad. Por otro lado, GPT-2 nos ha permitido generar texto coherente y creativo a partir de un "prompt" inicial, destacando su habilidad para continuar narrativas y construir frases significativas. Estos ejemplos ilustran el poder de los modelos de Transformer en diversas tareas de NLP, desde la comprensión hasta la creación de lenguaje.

<hr/>
<footer style="text-align:center; font-size:12px; color:gray;">
© 2026 UNAM Facultado de Ciencias – Todos los derechos reservados

</footer>